Cargo.toml

```diff
@@ -132,6 +132,14 @@ jj-lib = { path = "lib", version = "0.22.0" }
 jj-lib-proc-macros = { path = "lib/proc-macros", version = "0.22.0" }
 testutils = { path = "lib/testutils" }
 
+[workspace.lints.clippy]
+explicit_iter_loop = "deny"
+flat_map_option = "deny"
+implicit_clone = "deny"
+needless_for_each = "deny"
+semicolon_if_nothing_returned = "deny"
+uninlined_format_args = "deny"
```

Comment:
Perhaps, these rules should be all `"warn"`ed? It doesn't make sense that the default rules are warned, whereas extra rules are denied.

Reply:
Warnings are denied in the CI, or they would appear green:

```
cargo +stable clippy --all-features --workspace --all-targets -- -D warnings
```

It would rather make sense do deny them as well during development. I've added two patches to the serie:

- One near the beginning, which allows `clippy::too_long_first_doc_paragraph` with a higher priority, to avoid being bothered by this one until we have fixed the doc comments.
- One at the end, which sets the default clippy severity to deny, to match what is done in the CI.

Reply:
> It would rather make sense do deny them as well during development.

I disagree. Leaving e.g. unused variables should be okay when testing some ideas. It would be annoying if compilation failed because of minor nits.

Reply:
I was thinking the same, but I thought these checks would be done by `cargo clippy` only. Will they also be done as part of the regular build/test?

Reply:
All those lints are for Clippy, they will not make the compilation fail in any way, only runs of `cargo clippy` that you would use only when your code is ready to be shared.

Reply:
Also note that I'm not opposed to making them all warnings if people prefer, as long as the CI deny warnings it will keep the codebase just as clean.

Now that I think more about it, it might make sense if people use `cargo clippy` as a linter in their IDE (instead of `cargo check`, I know that VScode [that I don't use] can be configured that way).

Reply:
It's technically correct. However, compile and lint errors are both highlighted as errors in editor, so I think their error levels are conceptually the same.

Reply:
Making them all warnings and just having CI promote warnings to errors sounds good to me too.

Reply:
I've made them warnings, the CI will turn them into errors.